### DISCUSSION TITLE
how to build plugin architecture using micronaut with osgi or pf4j

### DISCUSSION THREAD

**Discussion by man-chi** [micronaut-projects/micronaut-core#10645] [General]
I would like to build a plugin/microkernel architecture, similar to the vscode extension for web applications. should I use Micronaut alone or integrate Micronaut with OSGi / plugin4java (PF4J) as my plugin engine?

should I use Micronaut as a DI engine at outer layer but having OSGi as inner layer for managing the plugin lifecycle.

how can I build a plugin extension on top of Micronaut? 

Please let me know.

**Comment by loicmathieu** [comment]
We build a plugin mechanism for Kestra based on Micronaut introspection beans (`@Introspected`) to discover the plugins, a custom plugin registry and a custom classloader.

Relying on bean introspection from Micronaut was a bad idea as it tied a version of Micronaut of the runtime, to a version of Micronaut on the plugin, so each time we want to update Micronaut, we need to update it on our 70 plugins, even for minor version (no idea why but bean introspection are created specific to a version).

So my advice is to have your plugins not depend on anything from Micronaut, we are thinking of removing the usage of bean introspection and relying instead on classloader scanning or other mechanism (Jandex for ex).

You can have a look at our plugin implementation here: https://github.com/kestra-io/kestra/tree/develop/core/src/main/java/io/kestra/core/plugins

**Reply by man-chi** [reply]
would you recommend using "plugin framework 4 java" https://pf4j.org/  with Micronaut?

**Reply by loicmathieu** [reply]
Didn't know it so it depends if it fits your needs.

My main point is: don't depend of Micronaut inside your plugin, ideally your plugin should have 0 deps or only a single API/model deps with no transitive dependency, or you will have issue during update.

**Reply by auke-** [reply]
We use the [Micronaut Bean Import](https://docs.micronaut.io/latest/guide/#beanImport) mechanisms to load [jakarta.inject](https://jakarta.ee/specifications/dependency-injection/2.0/apidocs/jakarta/inject/inject) annotated beans from our internal libraries.

We actually use [Java SPI](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html) to discover the plugins with the help of [Google AutoService](https://github.com/google/auto/blob/main/service/README.md) and let Micronaut instantiate the classes and inject dependent beans.

This way our plugin / library projects don't have a Micronaut dependency at all, just a dependency on the Jakarta annotations.

The Micronaut Maven Plugin also supports [importing beans from dependencies](https://micronaut-projects.github.io/micronaut-maven-plugin/latest/examples/bean-import.html) with minimal configuration.

**Reply by dstepanov** [reply]
@loicmathieu A Minor version upgrade shouldn't require all introspected classes to be recompiled.

**Reply by loicmathieu** [reply]
Well, it does, we had the issue multiple time, when a plugin has a version that is not the exact match of our main application, it crashes (the exception is a strange one , I'll add it here next time we migrate).
